x-pack/plugin/identity-provider/qa/idp-rest-tests/src/test/java/org/elasticsearch/xpack/idp/IdentityProviderAuthenticationIT.java

```diff
@@ -0,0 +1,240 @@
+        assertThat(samlResponse, containsString("InResponseTo=\"" + requestId + "\""));
+        authenticateWithSamlResponse(samlResponse, requestId);
+    }
+
+    private void checkIndexDoc(SamlServiceProviderIndex.DocumentVersion docVersion) throws IOException {
+        final Request request = new Request("GET", SamlServiceProviderIndex.INDEX_NAME + "/_doc/" + docVersion.id);
+        final Response response = adminClient().performRequest(request);
+        final Map<String, Object> map = entityAsMap(response);
+        assertThat(map.get("_index"), equalTo(SamlServiceProviderIndex.INDEX_NAME));
+        assertThat(map.get("_id"), equalTo(docVersion.id));
+        assertThat(asLong(map.get("_seq_no")), equalTo(docVersion.seqNo));
+        assertThat(asLong(map.get("_primary_term")), equalTo(docVersion.primaryTerm));
+    }
+
+    private SamlServiceProviderIndex.DocumentVersion createServiceProvider(String entityId, Map<String, Object> body) throws IOException {
+        final Request request =
+            new Request("PUT", "/_idp/saml/sp/" + encode(entityId) + "?refresh=" + WriteRequest.RefreshPolicy.IMMEDIATE.getValue());
+        final String entity = Strings.toString(JsonXContent.contentBuilder().map(body));
+        request.setJsonEntity(entity);
+        final Response response = client().performRequest(request);
+        final Map<String, Object> map = entityAsMap(response);
+        assertThat(ObjectPath.eval("service_provider.entity_id", map), equalTo(entityId));
+        assertThat(ObjectPath.eval("service_provider.enabled", map), equalTo(true));
+
+        final Object docId = ObjectPath.eval("document._id", map);
+        final Object seqNo = ObjectPath.eval("document._seq_no", map);
+        final Object primaryTerm = ObjectPath.eval("document._primary_term", map);
+        assertThat(docId, instanceOf(String.class));
+        assertThat(seqNo, instanceOf(Number.class));
+        assertThat(primaryTerm, instanceOf(Number.class));
+        return new SamlServiceProviderIndex.DocumentVersion((String) docId, asLong(primaryTerm), asLong(seqNo));
+    }
+
+    private Map<String, Object> validateAuthnRequest(String entityId, String authnRequestQuery) throws Exception {
+        final Request request = new Request("POST", "/_idp/saml/validate");
+        request.setJsonEntity("{\"authn_request_query\":\"" + authnRequestQuery + "\"}");
+        final Response response = client().performRequest(request);
+        final Map<String, Object> map = entityAsMap(response);
+        assertThat(ObjectPath.eval("service_provider.entity_id", map), instanceOf(String.class));
+        assertThat(ObjectPath.eval("service_provider.entity_id", map), equalTo(entityId));
+        assertThat(ObjectPath.eval("authn_state", map), instanceOf(Map.class));
+        return ObjectPath.eval("authn_state", map);
+    }
+
+    private Tuple<String, String> generateSamlAuthnRequest(String realmName) throws Exception {
```

Comment:
Suggestion:
This could return `SamlPrepareAuthenticationResponse` instead rather than using a tuple where the caller need to know what `v1` and `v2` mean.